src/shared/dissect-image.c

```diff
@@ -1270,27 +1321,60 @@ static int verity_partition(
 #else
-                r = log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "activation of verity device with signature requested, but not supported by cryptsetup due to missing crypt_activate_by_signed_key()");
+                        r = log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "activation of verity device with signature requested, but not supported by cryptsetup due to missing crypt_activate_by_signed_key()");
 #endif
-        } else
-                r = crypt_activate_by_volume_key(cd, name, root_hash, root_hash_size, CRYPT_ACTIVATE_READONLY);
-        if (r < 0)
-                return r;
+                } else
+                        r = crypt_activate_by_volume_key(cd, name, root_hash, root_hash_size, CRYPT_ACTIVATE_READONLY);
+                if (!IN_SET(r, 0, -EEXIST, -ENODEV, -EINVAL))
+                        return r;
+                /* libdevmapper can return EINVAL when the device is already in the activation stage.
+                 * There's no way to distinguish this situation from a genuine error due to invalid
+                 * parameters, so immediately fallback to activating the device with a unique name.
+                 * Improvements in libcrypsetup can ensure this never happens: https://gitlab.com/cryptsetup/cryptsetup/-/merge_requests/96 */
+                if (r == -EINVAL && FLAGS_SET(flags, DISSECT_IMAGE_VERITY_SHARE))
+                        return verity_partition(m, v, root_hash, root_hash_size, verity_data, NULL, root_hash_sig ?: hash_sig_from_file, root_hash_sig_size, flags & ~DISSECT_IMAGE_VERITY_SHARE, d);
+                if (r == -EEXIST) {
+                        if (!restore_deferred_remove){
+                                /* To avoid races, disable automatic removal on umount while setting up the new device. Restore it on failure. */
+                                r = dm_deferred_remove_cancel(name);
+                                if (r < 0)
+                                        return log_error_errno(r, "Disabling automated deferred removal for verity device %s failed: %m", node);
+                                restore_deferred_remove = strdup(name);
+                                if (!restore_deferred_remove)
+                                        return log_oom();
+                        }
+
+                        r = verity_can_reuse(root_hash, root_hash_size, !!root_hash_sig || !!hash_sig_from_file, name, &cd);
+                        if (!IN_SET(r, 0, -ENODEV, -ENOENT)) {
+                                return log_error_errno(r, "Checking whether existing verity device %s can be reused failed: %m", node);
+                        }
+                }
+                if (r == 0) {
+                        /* Sanity check: libdevmapper is known to report that the device already exists and is active,
+                         * but it's actually not there, so the later filesystem probe or mount would fail. */
+                        r = access(node, F_OK);
```

Comment:
nothing uses the return value here?

Reply:
r is set, loop is exited through break and r is checked immediately after

Reply:
I'll move this check outside of the loop, it will make it clearer what the intention is